python/strings_udf/cpp/include/cudf/strings/udf/udf_string.cuh

```diff
@@ -0,0 +1,493 @@
+  auto const nbytes     = m_bytes + char_bytes;
+  if (nbytes > m_capacity) { reallocate(2 * nbytes); }
+  auto out_ptr = m_data + m_bytes;
+  for (auto idx = 0; idx < count; ++idx) {
+    out_ptr += cudf::strings::detail::from_char_utf8(chr, out_ptr);
+  }
+  m_bytes         = nbytes;
+  m_data[m_bytes] = '\0';
+  return *this;
+}
+
+__device__ inline udf_string& udf_string::append(cudf::string_view const in)
+{
+  return append(in.data(), in.size_bytes());
+}
+
+__device__ inline udf_string& udf_string::operator+=(cudf::string_view const in)
+{
+  return append(in);
+}
+
+__device__ inline udf_string& udf_string::operator+=(cudf::char_utf8 chr) { return append(chr); }
+
+__device__ inline udf_string& udf_string::operator+=(char const* str) { return append(str); }
+
+__device__ inline udf_string& udf_string::insert(cudf::size_type pos,
+                                                 char const* str,
+                                                 cudf::size_type in_bytes)
+{
+  return replace(pos, 0, str, in_bytes);
+}
+
+__device__ inline udf_string& udf_string::insert(cudf::size_type pos, char const* str)
+{
+  return insert(pos, str, detail::bytes_in_null_terminated_string(str));
+}
+
+__device__ inline udf_string& udf_string::insert(cudf::size_type pos, cudf::string_view const in)
+{
+  return insert(pos, in.data(), in.size_bytes());
+}
+
+__device__ inline udf_string& udf_string::insert(cudf::size_type pos,
+                                                 cudf::size_type count,
+                                                 cudf::char_utf8 chr)
+{
+  return replace(pos, 0, count, chr);
+}
+
+__device__ inline udf_string udf_string::substr(cudf::size_type pos, cudf::size_type count) const
+{
+  if (pos < 0) { return udf_string{"", 0}; }
+  auto const spos = byte_offset(pos);
+  if (spos >= m_bytes) { return udf_string{"", 0}; }
+  auto const epos = count < 0 ? m_bytes : std::min(byte_offset(pos + count), m_bytes);
+  return udf_string{data() + spos, epos - spos};
+}
+
+// utility for replace()
+__device__ void udf_string::shift_bytes(cudf::size_type spos,
+                                        cudf::size_type epos,
+                                        cudf::size_type nbytes)
+{
+  if (nbytes < m_bytes) {
+    // shift bytes to the left [...wxyz] -> [wxyzxyz]
+    auto src = epos;
+    auto tgt = spos;
+    while (tgt < nbytes) {
+      m_data[tgt++] = m_data[src++];
+    }
+  } else if (nbytes > m_bytes) {
+    // shift bytes to the right [abcd...] -> [abcabcd]
+    auto src = m_bytes;
+    auto tgt = nbytes;
+    while (src > epos) {
+      m_data[--tgt] = m_data[--src];
+    }
+  }
+}
+
+__device__ inline udf_string& udf_string::replace(cudf::size_type pos,
+                                                  cudf::size_type count,
+                                                  char const* str,
+                                                  cudf::size_type in_bytes)
```

Comment:
Very minor nit: I just noticed that in the declaration in `udf_string.hpp` this argument is called `bytes` whereas here it is called `in_bytes`.

Reply:
This is allowed. It was likely intentional to keep the declaration consistent with the other functions but make it clearer in the definition. If you found that confusing then perhaps it is not worth it.